M2/Macaulay2/m2/methods.m2

```diff
@@ -1,159 +1,262 @@
+
 --		Copyright 1993-2002 by Daniel R. Grayson
+-*
+  Methods are the most common type of dynamic dispatch, along with Hooks.
+  The method function is stored under the youngest mutable hash table in the
+  method key sequence.
+*-
 
 needs "option.m2"
 
-MethodFunction = new Type of CompiledFunctionClosure
-MethodFunctionSingle = new Type of CompiledFunctionClosure
-MethodFunctionBinary = new Type of CompiledFunctionClosure
-MethodFunctionWithOptions = new Type of FunctionClosure
+-----------------------------------------------------------------------------
+-- Local variables
+-----------------------------------------------------------------------------
 
-MethodFunction.synonym = "method function"
-MethodFunctionSingle.synonym = "method function with a single argument"
-MethodFunctionBinary.synonym = "binary method function"
-MethodFunctionWithOptions.synonym = "method function with options"
+-----------------------------------------------------------------------------
+-- Local utilities
+-----------------------------------------------------------------------------
 
-dispatcherFunctions = {}
+-- see lists.m2
+all' := (L, p) -> not any(L, x -> not p x)
 
-noapp := (f,x) -> error(
-     "no method for applying item of class ", toString class f, 
-     " to item of class ", toString class x
-     )
+noapp := (f, x) -> concatenate("no method found for applying item of class ", toString class f, " to item of class ", toString class x)
+line0 :=  M     -> concatenate("no method found for applying ", silentRobustString(45, 3, M), " to:");
+
+printarg := (i, arg, out) -> horizontalJoin("     argument ", i, " :  ",
+    (if out then silentRobustNet else silentRobustNetWithClass)(60, 5, 3, arg));
 
-f := (i,arg,out) -> horizontalJoin("     argument ",i," :  ", (if out then silentRobustNet else silentRobustNetWithClass)(60,5,3,arg));
-line0 := meth -> concatenate("no method found for applying ", silentRobustString(45,3,meth), " to:");
+noMethodSingle = (M, args, outputs) -> toString stack     (  line0 M, printarg(" ", args, outputs) )
+noMethod       = (M, args, outputs) -> toString stack join( {line0 M},
+    if class args === Sequence and 0 < #args and #args <= 4 then apply(#args,
+	i -> printarg(toString (i+1), args#i, if outputs#?i then outputs#i else false))
+    else {   printarg(" ",            args,   false) }) -- TODO: do better here, in what way?
 
-noMethodSingle = (meth,args,output) -> error toString stack( line0 meth, f(" ",args,output))
-noMethod = (meth,args,outputs) -> error toString stack join( {line0 meth},
-     if class args === Sequence and 0 < #args and #args <= 4
-     then apply(#args, i -> f(toString (i+1),args#i,if outputs#?i then outputs#i else false))
-     else {f(" ",args,
-	       false					    -- do better here
-	       )})
+-- TODO: what does this for exactly?
```

Comment:
"does this for"??

Reply:
Should be "is for". What is `badClass` for?

Reply:
It's used further down in the file to print error messages.

Reply:
I can also use the search function! Can you show me a sample code that results in that error message? Specifically, this string:
```
expected an output class, but got:
```

Reply:
The comment is fixed, but I'll leave this conversation open as I'm still curious when that error can occur!

Reply:
Is this dead code?

Reply:
Probably not.

Reply:
Then can you give me an example of code that results in that error?

Reply:
Not without a bit of study ...

Reply:
Could you do a bit of study then? I spent quite a bit trying to figure this out and tracked it down to the `env.2` argument here:
https://github.com/Macaulay2/M2/blob/3c3e3f7ddbc1a0ff20bd6723e41251557dea2936/M2/Macaulay2/d/actors5.d#L800-L804
but this comment doesn't make sense and the code below it isn't much better either. Is it for making a method based on the output argument? Does M2 support such a thing? It would really help if you documented your code in the `d` directory.

Reply:
Determining whether 1 line of code is dead doesn't seem urgent enough.

Reply:
I asked this 26 days ago. When do you think you would get around to it?

Reply:
Never, for even if one could prove that there is no way to trigger the error message from top level currently, it might be worthwhile to keep the error-detecting code present, as a way of preventing "internal" errors, caused by future changes to the code.  Such error-detection is especially valuable when chunks of code in widely separate areas must be closely coordinated.

Reply:
If you have an error that can't be triggered it gives the false belief that the code is safe, when in reality it may not be. Also keeping dead code around is bad etiquette for future readers and maintainers, so I'll delete it when I simplify the `d` code for methods in my next PR.

Reply:
No, let it stay.

Reply:
The way I'm simplifying `newmethod1234c` there's no point to passing this `badClass` function as an argument. If I'm misunderstanding something about `newmethod1234c`, please let me know, though it seems like you prefer that I don't wait.

Reply:
Is there any reason to change `newmethod1234c`?

Reply:
Yes. The methods it creates are slow when options are given. It's a major bottleneck for me.

Reply:
How many times do you have to call the methods to perform your computation?

Do you have some code that demonstrates the slowness?

Reply:
Sure, take a look at the benchmarks in `tests/normal/methods.m2`.

Reply:
What benchmarks?  (I don't see any timing commands or any use of the Benchmark package there.)

Reply:
If you don't see the tests I've added then I'm sorry but I really don't have the time or energy to keep debating you and convincing you that the code you haven't touched in more than a decade is in need of improvement, especially when you refuse to explain your code when I ask about it. Try writing any of your own benchmarks and compare with other symbolic algebra software.